### ISSUE TITLE
Trigger onAuthStateChange after setSession

### ISSUE THREAD

**Issue by iveshenry18** [supabase/gotrue-js#588]
# Feature request
Currently, calling `setSession` does not trigger the callback(s) registered in `onAuthStateChange`. I think it should!

## Is your feature request related to a problem? Please describe.
It came as a surprise to me, given the naming and documentation of `onAuthStateChange`, that the callbacks therein are not triggered when running `setSession`. Semantically, "setting the session" seems to fall under the category of an "auth state change".

I am building an application with Supabase with Expo. I am using Expo's flow to kick off SSO via Supabase, then passing the `access_token` and `refresh_token` to `supabase.auth.setSession`. On the React side, I am using the common pattern of storing the session as top-level state, then registering a setState call in `onAuthStateChange` (within a `useEffect`). I expected this to work as-is, but found that the setState call was not getting triggered by the `supabase.auth.setSession` call, preventing the necessary re-render after the user completes the login flow.

As a temporary workaround, I call `refreshSession` after `setSession`, which _does_ trigger the `onAuthStateChange` callback (the `setState`). It's a bit hacky, but seems to be working for now.

```ts
const continueWithProvider = async (provider: string) => {
  const redirectUrl = makeRedirectUri({ path: "/auth/callback" });

  // Trigger Supabase auth flow via Expo
  const authResponse = await startAsync({
    authUrl: `${supabaseApiUrl}/auth/v1/authorize?provider=${provider}&redirect_to=${redirectUrl}`,
  });

  if (authResponse.type === "success") {
    try {
      // Use credentials from Supabase-via-Expo auth flow to set Supabase session
      // I would expect this call to trigger the supabase.auth.onAuthStateChange callbacks
      const { data: { session } } = await supabase.auth.setSession({
        access_token: authResponse.params.access_token,
        refresh_token: authResponse.params.refresh_token,
      });

      // I added this call to "force" the onAuthStateChange callback(s)
      supabase.auth.refreshSession();
    } catch (err) {
      console.warn(`Unable to set Supabase session`);
      console.warn(err);
      throw err;
    }

```

```ts
export default function useAuth() {
  const [session, setSession] = useState<Session | null>(null);

  useEffect(() => {
    supabase.auth.getSession().then(({ data: { session } }) => {
      setSession(session);
    });

    // I would expect this to be called after running supabase.auth.setSession()
    const { data: authListener } = supabase.auth.onAuthStateChange(
      async (_event, session) => {
        setSession(session);
      }
    );
    return () => {
      if (authListener) authListener.subscription.unsubscribe();
    };
  }, [supabase.auth]);

  return session
}
```

## Describe the solution you'd like

Unless there is a good reason not to, I think `setSession` should call `_notifyAllSubscribers`. 

## Describe alternatives you've considered

As mentioned, I have a workaround, but it's a bit hacky. The other option is to copy any `onAuthStateChange` callbacks and just call them directly after running `setSession`, but that seems to defeat the purpose of `onAuthStateChange`.

## Additional context

Classic caveat: I may not have any clue what I'm talking about. Feel free to let me know if that's the case. Thanks!

**Comment by j4w8n** [comment]
This functionality was introduced with gotrue-js 2.8.0. They haven't bumped the version in supabase-js yet, but they likely will next week.

https://github.com/supabase/gotrue-js/pull/581

**Comment by iveshenry18** [comment]
Great, thank you!

**Comment by j4w8n** [comment]
This has been implemented, in supabase-js v2.6.0 for sure.
